pilot/pkg/config/mcp/controller.go

```diff
@@ -0,0 +1,248 @@
+	}
+
+	ns, exists := c.configStore[typ][namespace]
+	if !exists {
+		c.logger.Infof("Get: config not found for the type %s", typ)
+		return nil, false
+	}
+
+	out, exists := ns.Load(name)
+	if !exists {
+		return nil, false
+	}
+	config := out.(model.Config)
+
+	return &config, true
+}
+
+func (c *Controller) List(typ, namespace string) ([]model.Config, error) {
+	_, ok := c.ConfigDescriptor().GetByType(typ)
+	if !ok {
+		return nil, errors.New(fmt.Sprintf("List: unknown type %s", typ))
+	}
+	data, exists := c.configStore[typ]
+	if !exists {
+		c.logger.Infof("List: config not found for the type %s", typ)
+		return nil, nil
+	}
+	out := make([]model.Config, 0, len(c.configStore[typ]))
+	if namespace == "" {
+		for _, ns := range data {
+			ns.Range(func(key, value interface{}) bool {
+				out = append(out, value.(model.Config))
+				return true
+			})
+		}
+	} else {
+		ns, exists := data[namespace]
+		if !exists {
+			return nil, nil
+		}
+		ns.Range(func(key, value interface{}) bool {
+			out = append(out, value.(model.Config))
+			return true
+		})
+	}
+	return out, nil
+}
+
+func (c *Controller) Update(config model.Config) (newRevision string, err error) {
+	typ := config.Type
+	schema, ok := c.ConfigDescriptor().GetByType(typ)
+	if !ok {
+		return "", errors.New(fmt.Sprintf("Update: unknown type %s", typ))
+	}
+	if err := schema.Validate(config.Name, config.Namespace, config.Spec); err != nil {
+		return "", err
+	}
+
+	ns, exists := c.configStore[typ][config.Namespace]
+	if !exists {
+		return "", errNotFound
+	}
+
+	oldConfig, exists := ns.Load(config.Name)
+	if !exists {
+		return "", errNotFound
+	}
+
+	if config.ResourceVersion != oldConfig.(model.Config).ResourceVersion {
+		return "", errors.New("old revision")
+	}
+
+	rev := time.Now().String()
+	config.ResourceVersion = rev
+	ns.Store(config.Name, config)
+	return rev, nil
+}
+
+func (c *Controller) Create(config model.Config) (revision string, err error) {
+	typ := config.Type
+	schema, ok := c.ConfigDescriptor().GetByType(typ)
+	if !ok {
+		return "", errors.New("unknown type")
+	}
+	if err := schema.Validate(config.Name, config.Namespace, config.Spec); err != nil {
+		return "", err
+	}
+	ns, exists := c.configStore[typ][config.Namespace]
+	if !exists {
+		ns = new(sync.Map)
+		c.configStore[typ][config.Namespace] = ns
+	}
+
+	_, exists = ns.Load(config.Name)
+
+	if !exists {
+		tnow := time.Now()
+		config.ResourceVersion = tnow.String()
+
+		// Set the creation timestamp, if not provided.
+		if config.CreationTimestamp.Time.IsZero() {
+			config.CreationTimestamp.Time = tnow
+		}
+
+		ns.Store(config.Name, config)
+		return config.ResourceVersion, nil
+	}
+	return "", errAlreadyExists
+}
+
+func (c *Controller) Apply(change *mcpclient.Change) error {
+	for _, obj := range change.Objects {
+		descriptor, ok := c.descriptorsByMessageName[change.MessageName]
+		if !ok {
+			return fmt.Errorf("Apply: type not supported %s", change.MessageName)
+		}
+
+		c.updateCh <- changeEvent{
```

Comment:
What is the rationale for populating the configStore via a channel vs. writing it directly?

Reply:
Good point I will update this since sync.Map seems adequate.